Add harvester network deep dive document (#231)
harvester/docs: docs/networking/deep-dive.md

---
sidebar_position: 3
sidebar_label: Deep Dive
title: ""
keywords:
- Harvester
- Networking
- Topology
---
# Harvester Network Deep Dive

The network topology below reveals how we implement the Harvester network.

![](/img/v1.1/networking/topology.png)

> The diagram contains [the built-in cluster network mgmt](./clusternetwork.md/#built-in-cluster-network) and a [custom cluster network](./clusternetwork.md/#custom-cluster-network) called `oob`.

Reviewer: ```suggestion
The diagram contains [the built-in cluster network mgmt](./clusternetwork.md/#built-in-cluster-network) and a [custom cluster network](./clusternetwork.md/#custom-cluster-network) called `OOB`.
```

Author: In the diagram, the bridge name is `oob-br` and the bond name is `oob-bo`, so the cluster network is exactly `oob`, not `OOB`.

As shown above, the Harvester network focuses on layer 2. We leverage Linux network devices to construct traffic paths for the communication between VM and VM, VM and host, and VM and external network devices.

The Harvester network is divided into three layers, including:

- kubevirt networking layer

- Harvester networking layer

- external networking layer

## Kubevirt Networking

> The basic idea of kubevirt is to put libvirt into the Kubernetes pod. The kubevirt networking is responsible for forwarding the traffic between the pod and the VM inside.
> Please refer to the [kubevirt official document](https://kubevirt.io/2018/KubeVirt-Network-Deep-Dive.html) for more details.

Reviewer: ```suggestion
The basic idea of KubeVirt is to put libvirt into the Kubernetes pod. The KubeVirt network is responsible for forwarding traffic between the pod and VM. 
Please refer to the [KubeVirt official document](https://kubevirt.io/2018/KubeVirt-Network-Deep-Dive.html) for more details.
```

## Harvester Networking

> Harvester networking aims to forward the traffic from pods to the host network. It implements the management network, VLAN(tagged) network and untagged network. We temporarily refer to the latter two as **bridge network**, because bridge plays a key role in their implementation.

Reviewer: ```suggestion
Harvester networking is designed to forward traffic from pods to the host network, It implements a management network, VLAN networks and untagged networks. We can refer to the last two networks as bridged networks because bridging plays a vital role in their implementation.
```

Author: The word `bridged` looks strange.

Author: I prefer to keep the name `bridge network`.

### Bridge Network

We leverage [multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni) and [bridge CNI](https://www.cni.dev/plugins/current/main/bridge/) to implement the bridge network.

> 1. Multus CNI is a container network interface (CNI) plugin for Kubernetes that enables attaching multiple network interfaces to pods. Its capability allows a VM to be equipped with both a NIC for the management network and multiple NICs for the bridge network. Without multus CNI, the bridge CNI will not appear in Harvester.

Reviewer: ```suggestion
1. Multus CNI is a Container Network Interface (CNI) plugin for Kubernetes that can attach multiple network interfaces to a pod. Its capability allows a VM to have one NIC for the management network and multiple NICs for the bridged network.
```

> 2. With bridge CNI, the VM pods will be plugged into an L2 bridge specified in the network attachment definition config.

Reviewer: ```suggestion
2. Using the bridge CNI, the VM pod will be plugged into the L2 bridge specified in the Network Attachment Definition config.
```

    ```json
    # Example 1
    {
        "cniVersion": "0.3.1",
        "name": "vlan100",
        "type": "bridge",
        "bridge": "mgmt-br",
        "promiscMode": true,
        "vlan": 100,
    }
    ```

    ```json
    # Example 2
    {
        "cniVersion": "0.3.1",
        "name": "untagged-network",
        "type": "bridge",
        "bridge": "oob-br",
        "promiscMode": true,
        "ipam": {}
    }
    ```

>     Example 1 is a typical VLAN configuration whose VLAN ID is 100, while example 2 is an untagged network configuration without a VLAN ID. The VM pods using example 1 configuration will be plugged into the bridge mgmt-br, while those using example 2 will be plugged into the bridge oob-br.

Reviewer: ```suggestion
    Example 1 is a typical VLAN configuration with VLAN ID 100, while Example 2 is an untagged network configuration with no VLAN ID. The VM pod configured using Example 1 will be plugged into the bridge `mgmt-br`, while the VM pod using Example 2 will be plugged into the bridge `oob-br`.
```

> 3. For fault tolerance and high availability, the uplink of every bridge is a bond device. As a bridge port, the uplink should allow the tagged package to pass.

Reviewer: ```suggestion
3. For fault tolerance and high availability concerns, the uplink of each bridge is a bond device. For a bridge port, the uplink should allow the tagged package to pass directly.
```

    ```shell
    harvester-0:/home/rancher # bridge -c vlan show dev oob-bo
    port	   vlan ids
    oob-bo	   1 PVID Egress Untagged
               100
               200
    ```

    The example above shows that the bond `oob-bo` allows packages with tag 1, 100 or 200.

### Management Network

> The management network is based on [canal](https://projectcalico.docs.tigera.io/getting-started/kubernetes/flannel/flannel).

Reviewer: ```suggestion
The management network is based on [Canal](https://projectcalico.docs.tigera.io/getting-started/kubernetes/flannel/flannel).
```

> It is worth mentioning that the canal interface is the bridge `mgmt-br` or the VLAN sub-interface of `mgmt-br`, where Harvester configures the node IP. This design has two benefits:
> 
> - The built-in cluster network mgmt supports the management network and bridge network simultaneously.
> - Through a VLAN network interface, we can specify a VLAN ID to the management network.

Reviewer: ```suggestion
It is worth mentioning that the Canal interface uses the bridge `mgmt-br` and the VLAN sub-interface of `mgmt-br` where the Harvester configures its node IP. This design has two benefits:

- The built-in `mgmt` cluster network supports both the management network and bridge network.
- With the VLAN network interface, we can assign a VLAN ID to the management network.
```

> - Through a VLAN network interface, we can specify a VLAN ID to the management network.

Author: I think `to` is more accurate.

## External Networking

> The external network devices usually include switches and a DHCP server. After introducing the cluster network, we can group the host's NICs and connect them to different switches for traffic isolation. Below are some usage notes.

Reviewer: ```suggestion
External network devices typically refer to switches and DHCP servers. With a cluster network, we can group host NICs and connect them to different switches for traffic isolation. Below are some usage instructions.
```

> - To allow the passing of the tagged packages, you need to set the type of external switch ports connected to the hosts or other devices (such as the DHCP server) as trunk or hybrid and permit the specified VLANs.

Reviewer: ```suggestion
- To allow tagged packets to pass, you need to set the port type of the external switch or other devices (such as a DHCP server) to trunk or hybrid mode and allow the specified VLAN tag.
```

> - To allow the passing of the tagged packages, the external switch ports connected to the hosts or other devices (such as the DHCP server) should be set as trunk or hybrid type and permit the specified VLANs.

Reviewer 2: ```suggestion
- To allow the passing of the tagged packages, you should set the link type of external switch ports connected to the hosts or other devices (such as the DHCP server) to trunk or hybrid and permit the specified VLANs.
```

> - We need to configure link aggregation on the switch based on the bond mode of the peer host.

Reviewer: can we add an example of this?

Author: List the correspondence between bond mode and link aggregation mode. And we will provide examples in the user case document. cc @tjjh89017

> - If you want the VMs in a VLAN to be able to obtain an IP address via the DHCP protocol, configure an IP pool for that VLAN in the DHCP server.

Reviewer: ```suggestion
- If you want VMs in a VLAN to be able to obtain IP addresses through the DHCP protocol, configure an IP pool for that VLAN in the DHCP server.
```

---
sidebar_position: 3
sidebar_label: Deep Dive
title: ""
keywords:
- Harvester
- Networking
- Topology
---
# Harvester Network Deep Dive

The network topology below reveals how we implement the Harvester network.

![](/img/v1.1/networking/topology.png)

The diagram contains [the built-in cluster network mgmt](./clusternetwork.md/#built-in-cluster-network) and a [custom cluster network](./clusternetwork.md/#custom-cluster-network) called `oob`.

> As shown above, the Harvester network primarily focuses on layer 2. We leverage Linux network devices to construct traffic paths for the communication between VM to VM, VM to host, and VM to external network devices.

Reviewer 3: `layer 2` -> `OSI model layer 2`
`Linux network devices` -> `Linux network devices and protocols`

> The Harvester network is divided into three layers, including:

Reviewer 3: `is divided into three layers` -> `is composed of multi layers`

- KubeVirt networking layer

- Harvester networking layer

- external networking layer

## Kubevirt Networking

> The basic idea of KubeVirt is to put libvirt into the Kubernetes pod. The KubeVirt network is responsible for forwarding traffic between the pod and VM. 

Reviewer 3: `The basic idea of KubeVirt is to put libvirt into the Kubernetes pod.`  -> `The general purpose of KubeVirt is to run VM inside the Kubernetes pod. `

`The KubeVirt network is responsible for forwarding traffic between the pod and VM. ` -> `The KubeVirt network builds the network path between the pod and VM. `

Please refer to the [KubeVirt official document](https://kubevirt.io/2018/KubeVirt-Network-Deep-Dive.html) for more details.

## Harvester Networking

> Harvester networking is designed to forward the traffic from pods to the host network. It implements a management network, VLAN networks and untagged networks. We can refer to the last two networks as **bridge networks**, because bridge plays a vital role in their implementation.

Reviewer 3: strictly speaking, `Harvester networking is designed to forward the traffic from pods to the host network.` is `Harvester networking is designed to forward the traffic from pods to the host network and vice versa.`

should we use `forward the traffic from .. to...` ? or `build/construct the network (path) between ... and ...` ?  network is not only `forward  the traffic`

1. Multus CNI is a Container Network Interface (CNI) plugin for Kubernetes that can attach multiple network interfaces to a pod. Its capability allows a VM to have one NIC for the management network and multiple NICs for the bridge network.

2. Using the bridge CNI, the VM pod will be plugged into the L2 bridge specified in the Network Attachment Definition config.

    ```json
    # Example 1
    {
        "cniVersion": "0.3.1",
        "name": "vlan100",
        "type": "bridge",
        "bridge": "mgmt-br",
        "promiscMode": true,
        "vlan": 100,
    }
    ```

    ```json
    # Example 2
    {
        "cniVersion": "0.3.1",
        "name": "untagged-network",
        "type": "bridge",
        "bridge": "oob-br",
        "promiscMode": true,
        "ipam": {}
    }
    ```

    Example 1 is a typical VLAN configuration with VLAN ID 100, while Example 2 is an untagged network configuration with no VLAN ID. The VM pod configured using Example 1 will be plugged into the bridge `mgmt-br`, while the VM pod using Example 2 will be plugged into the bridge `oob-br`.

> 3. For fault tolerance and high availability concerns, the uplink of each bridge is a bond device. For a bridge port, the uplink should allow the tagged package to pass directly.

Reviewer 3: `For fault tolerance and high availability concerns, the uplink of each bridge is a bond device` -> `To achieve high availability and fault tolerance, a bond device is created to serve as the uplink of the bridge, all of the real uplink NICs are binded into this bond device.`

`For a bridge port, the uplink should allow the tagged package to pass directly.` -> `By default, this bond device will allow the target tagged traffic/packet to pass through`.

    ```shell
    harvester-0:/home/rancher # bridge -c vlan show dev oob-bo
    port	   vlan ids
    oob-bo	   1 PVID Egress Untagged
               100
               200
    ```

    The example above shows that the bond `oob-bo` allows packages with tag 1, 100 or 200.

### Management Network

The management network is based on [Canal](https://projectcalico.docs.tigera.io/getting-started/kubernetes/flannel/flannel).

> 

Reviewer 3: For Harvester `management network`, could more description be added, or reference to exising parts?

From L#85 to L#87, it jumps.

Author: @tjjh89017 Do you want to add more details about the management network?

Reviewer 4: It's fine for me.

> It is worth mentioning that the Canal interface uses the bridge `mgmt-br` and the VLAN sub-interface of `mgmt-br` where the Harvester configures its node IP. This design has two benefits:

Reviewer 3: is `mgmt-br` created by default and can not be deleted? when true, please highlight it here or in the previous part

- The built-in `mgmt` cluster network supports both the management network and bridge network.
- With the VLAN network interface, we can assign a VLAN ID to the management network.

## External Networking

External network devices typically refer to switches and DHCP servers. With a cluster network, we can group host NICs and connect them to different switches for traffic isolation. Below are some usage instructions.

- To allow tagged packets to pass, you need to set the port type of the external switch or other devices (such as a DHCP server) to trunk or hybrid mode and allow the specified VLAN tag.

> - We need to configure link aggregation on the switch based on the bond mode of the peer host. Link aggregation can work in manual mode or LACP mode. The following lists the correspondence between bond mode and link aggregation mode.

Reviewer 3: We -> You

---
sidebar_position: 3
sidebar_label: Deep Dive
title: ""
keywords:
- Harvester
- Networking
- Topology
---
# Harvester Network Deep Dive

The network topology below reveals how we implement the Harvester network.

![](/img/v1.1/networking/topology.png)

The diagram contains [the built-in cluster network mgmt](./clusternetwork.md/#built-in-cluster-network) and a [custom cluster network](./clusternetwork.md/#custom-cluster-network) called `oob`.

As shown above, the Harvester network primarily focuses on OSI model layer 2. We leverage Linux network devices and protocols to construct traffic paths for the communication between VM to VM, VM to host, and VM to external network devices.

The Harvester network is composed of three layers, including:

- KubeVirt networking layer

- Harvester networking layer

- external networking layer

> ## Kubevirt Networking

Reviewer 5: ```suggestion
## KubeVirt Networking
```
nit.

- The built-in `mgmt` cluster network supports both the management network and bridge network.
- With the VLAN network interface, we can assign a VLAN ID to the management network.

As components of the mgmt cluster network, it's not allowed to delete or modify the bridge mgmt-br, the bond mgmt-bo and the VLAN device.

## External Networking

External network devices typically refer to switches and DHCP servers. With a cluster network, we can group host NICs and connect them to different switches for traffic isolation. Below are some usage instructions.

- To allow tagged packets to pass, you need to set the port type of the external switch or other devices (such as a DHCP server) to trunk or hybrid mode and allow the specified VLAN tag.

- You need to configure link aggregation on the switch based on the bond mode of the peer host. Link aggregation can work in manual mode or LACP mode. The following lists the correspondence between bond mode and link aggregation mode.

    | Bond Mode | Link Aggregation Mode |
    | --- | ----------- |

>     | mode 0(balance-rr) | manual |

Reviewer 4: Could you change "manual" to "static link aggregation" to clarify
